app/models/data_repo_client.rb

```diff
@@ -0,0 +1,199 @@
+        Rails.logger.error "Retry count exceeded when requesting '#{path}' - #{error_message}"
+        raise e
+      end
+    end
+  end
+
+  ##
+  # API ENDPOINT BINDINGS
+  ##
+
+  ##
+  # Datasets
+  ##
+
+  # fetch all available datasets
+  #
+  # * *params*
+  #   - +direction+ (String) => Sort direction (default: desc)
+  #   - +filter+ (String) => Filter the results where this string is a case insensitive match in the name or description.
+  #   - +limit+ (Integer) => The numbers datasets to retrieve
+  #   - +offset+ (Integer) => The number of datasets to skip before when retrieving the next page
+  #   - +sort+ (String) => The field to use for sorting: (name, description, created_date)
+  #
+  # * *returns*
+  #   - (Array<Hash>) => Array of dataset JSON attributes
+  def datasets(direction: 'desc', filter: nil, limit: 100, offset: 0, sort: 'name')
+    validate_argument(:direction, direction, SORT_DIRECTIONS)
+    validate_argument(:sort, sort, SORT_OPTIONS)
+    query_opts = merge_query_options({direction: direction, filter: filter, limit: limit, offset: offset, sort: sort})
+    path = api_root + '/api/repository/v1/datasets' + query_opts
+    process_api_request(:get, path)
+  end
+
+  # fetch a single dataset
+  #
+  # * *params*
+  #   - +dataset_id+ (String: UUID) => Dataset UUID
+  #   - +include+ (Array) => Dataset fields to return from DATASET_INCLUDE_FIELDS (default: SCHEMA,PROFILE,DATA_PROJECT,STORAGE)
+  #
+  # * *returns*
+  #   - (Hash) => Hash of dataset JSON attributes
+  def dataset(dataset_id, include: ALL_DATASET_FIELDS)
+    path = api_root + "/api/repository/v1/datasets/#{dataset_id}"
+    if include.any?
+      validate_argument(:include, include, DATASET_INCLUDE_FIELDS)
+      query_opts = include.map {|i| "include=#{i}"}.join('&')
+      path += "?#{query_opts}"
+    end
+    process_api_request(:get, path)
+  end
+
+  ##
+  # Snapshots
+  ##
+
+  # fetch all available snapshots
+  #
+  # * *params*
+  #   - +datasetIds+ (Array<String>) => Filter the results where these datasetIds are source datasets.
+  #   - +direction+ (String) => Sort direction (default: desc)
+  #   - +filter+ (String) => Filter the results where this string is a case insensitive match in the name or description.
+  #   - +limit+ (Integer) => The numbers datasets to retrieve
+  #   - +offset+ (Integer) => The number of datasets to skip before when retrieving the next page
+  #   - +sort+ (String) => The field to use for sorting: (name, description, created_date)
+  #
+  # * *returns*
+  #   - (Array<Hash>) => Array of snapshot JSON atributes
+  def snapshots(datasetIds: [], direction: 'desc', filter: nil, limit: 100, offset: 0, sort: 'name')
+    validate_argument(:direction, direction, SORT_DIRECTIONS)
+    validate_argument(:sort, sort, SORT_OPTIONS)
+    query_opts = merge_query_options(
+      {
+        datasetIds: datasetIds.join(','), direction: direction, filter: filter,
+        limit: limit, offset: offset, sort: sort
```

Comment:
Usually I want a Ruby language feature in JavaScript, but here it's vice versa.  [Object initializer shorthand](https://stackoverflow.com/q/37365394) would be useful, but TIL it isn't available in Ruby.

Reply:
TOTALLY agree.  I've thought that many times myself.